iree/hal/metal/metal_command_buffer.mm

```diff
@@ -0,0 +1,140 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/hal/metal/metal_command_buffer.h"
+
+#include "iree/base/status.h"
+#include "iree/base/tracing.h"
+
+namespace iree {
+namespace hal {
+namespace metal {
+
+// static
+StatusOr<ref_ptr<CommandBuffer>> MetalCommandBuffer::Create(
+    CommandBufferModeBitfield mode, CommandCategoryBitfield command_categories,
+    id<MTLCommandBuffer> command_buffer) {
+  return assign_ref(new MetalCommandBuffer(mode, command_categories, command_buffer));
+}
+
+MetalCommandBuffer::MetalCommandBuffer(CommandBufferModeBitfield mode,
+                                       CommandCategoryBitfield command_categories,
+                                       id<MTLCommandBuffer> command_buffer)
+    : CommandBuffer(mode, command_categories), metal_handle_([command_buffer retain]) {}
+
+MetalCommandBuffer::~MetalCommandBuffer() {
+  IREE_TRACE_SCOPE0("MetalCommandBuffer::dtor");
+  [metal_handle_ release];
+}
+
+Status MetalCommandBuffer::Begin() {
+  IREE_TRACE_SCOPE0("MetalCommandBuffer::Begin");
+  is_recording_ = true;
+  return OkStatus();
+}
+
+Status MetalCommandBuffer::End() {
+  IREE_TRACE_SCOPE0("MetalCommandBuffer::End");
+  is_recording_ = false;
+  return OkStatus();
```

Comment:
Hmm.. really nothing to do upon ending command encoding?
_Finishes reading through the rest of the file_
Oh. This is all unimplemented boilerplate :P 
Carry on...

Reply:
Haha, yup. We need to end the inflight command encoder. But that's gonna happen in a following pull request. :)